lib/convection/model/template.rb

```diff
@@ -183,9 +197,122 @@ def diff(other, stack_ = nil)
+          end
+        end
+        mapping_attributes = mappings.values.flat_map do |inner_hash|
+          inner_hash.keys.select do |key|
+            value = inner_hash[key]
+          end
+        end
+        mapping_attributes.each do |attribute|
+          if attribute.length > CF_MAX_MAPPING_ATTRIBUTE_NAME
+            raise ArgumentError, "Error: Excessive Mapping Attribute Name #{attribute}"
+          end
+        end
+      end
+
+      def validate_parameters(rendered_stack = nil)
+        parameters= rendered_stack['Parameters']
+        number_of_parameters = parameters.count
+        if number_of_parameters > CF_MAX_PARAMETERS
+          raise ArgumentError, "Error: Excessive Number of Parameters (#{number_of_parameters}) Max= #{CF_MAX_PARAMETERS}"
+        end
+        largest_parameter_name = parameters.keys.max
+        if largest_parameter_name == nil
+          largest_parameter_name = ""
+        end
+        parameter_name_characters = largest_parameter_name.length
+        if parameter_name_characters > CF_MAX_PARAMETER_NAME_CHARACTERS
+          raise ArgumentError, "Error: Parameter Name #{largest_parameter_name} has too many characters (#{parameter_name_characters}) Max= #{CF_MAX_PARAMETER_NAME_CHARACTERS}"
+        end
+        parameters.values.each do |value|
+          parameter_in_json = JSON.generate(value)
+          if parameter_in_json.bytesize > CF_MAX_PARAMETER_VALUE_BYTESIZE
+            raise ArgumentError, "Error: Excessive Parameter Size #{value}"
+          end
+        end
+      end
+
+      def validate_outputs(rendered_stack = nil)
+        outputs = rendered_stack['Outputs']
+        number_of_outputs = outputs.count
+        if number_of_outputs > CF_MAX_OUTPUTS
+          raise ArgumentError, "Error: Excessive Number of Outputs (#{number_of_outputs}) Max= #{CF_MAX_OUTPUTS}"
+        end
+        largest_output_name = outputs.keys.max
+        if largest_output_name == nil
+          largest_output_name = ""
+        end
+        output_name_characters = largest_output_name.length
+        if output_name_characters > CF_MAX_OUTPUT_NAME_CHARACTERS
+          raise ArgumentError, "Error: Output Name #{largest_output_name} has too many characters (#{output_name_characters}) Max= #{CF_MAX_OUTPUT_NAME_CHARACTERS}"
+        end
+      end
+
+      def validate_description(rendered_stack = nil)
+        description_bytesize = rendered_stack['Description'].bytesize
+        if description_bytesize > CF_MAX_DESCRIPTION_BYTESIZE
+          raise ArgumentError, "Error: Excessive Description Size (#{description_bytesize}) Max= #{CF_MAX_DESCRIPTION_BYTESIZE}"
+        end
+      end
+
+      def validate_bytesize(rendered_stack = nil)
+        json = JSON.generate(rendered_stack)
+        template_bytesize = json.bytesize
+        if template_bytesize > CF_MAX_BYTESIZE
+          raise ArgumentError, "Error: Excessive Template Size (#{template_bytesize}) Max= #{CF_MAX_BYTESIZE}"
+        end
+      end
+
+
+
```

Comment:
You could take this a step further and have something like:

```
  def check_value_is_within_limits(val, limit, description)
    raise ArgumentError, "Error: #{description} is #{val}; limit is #{limit}" if val > limit
  end
```

  You would call it as follows:

```
  check_value_is_within_limits(json.bytesize, CF_MAX_BYTESIZE, 'template size')
```

This should look a bit cleaner, and maybe get rid of a few local variables, too.